### ISSUE TITLE
twisted.test.test_process.ProcessTestCase.testManyProcesses fails on WinXP

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3404]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#3404|
|Type|defect|
|Created|2008-09-01 12:28:11Z|
|Branch|https://github.com/twisted/twisted/tree/manyprocesses-3404-5|

This is consistent now.  It's not really clear when it was introduced.  It may have been the result of a configuration change on the host, rather than a code change, but I'm not certain.

```
===============================================================================
[ERROR]: twisted.test.test_process.ProcessTestCase.testManyProcesses

Traceback (most recent call last):
Failure: twisted.internet.defer.TimeoutError: <twisted.test.test_process.ProcessTestCase testMethod=testManyProcesses> (testManyProcesses) still running at 120.0 secs
===============================================================================
[ERROR]: twisted.test.test_process.ProcessTestCase.testManyProcesses

Traceback (most recent call last):
  File "c:\twistedbuildbot24\WXP32-full2.5-scmikes-select\Twisted\twisted\internet\base.py", line 757, in runUntilCurrent
    call.func(*call.args, **call.kw)
  File "c:\twistedbuildbot24\WXP32-full2.5-scmikes-select\Twisted\twisted\internet\_pollingfile.py", line 72, in _pollEvent
    workUnits += resource.checkWork()
  File "c:\twistedbuildbot24\WXP32-full2.5-scmikes-select\Twisted\twisted\internet\_pollingfile.py", line 129, in checkWork
    self.cleanup()
  File "c:\twistedbuildbot24\WXP32-full2.5-scmikes-select\Twisted\twisted\internet\_pollingfile.py", line 134, in cleanup
    self.lostCallback()
  File "c:\twistedbuildbot24\WXP32-full2.5-scmikes-select\Twisted\twisted\internet\_dumbwin32proc.py", line 289, in outConnectionLost
    self.proto.childConnectionLost(1)
  File "c:\twistedbuildbot24\WXP32-full2.5-scmikes-select\Twisted\twisted\test\test_process.py", line 205, in childConnectionLost
    raise RuntimeError
exceptions.RuntimeError: 
===============================================================================
[ERROR]: twisted.test.test_process.ProcessTestCase.testManyProcesses

Traceback (most recent call last):
  File "c:\twistedbuildbot24\WXP32-full2.5-scmikes-select\Twisted\twisted\internet\base.py", line 757, in runUntilCurrent
    call.func(*call.args, **call.kw)
  File "c:\twistedbuildbot24\WXP32-full2.5-scmikes-select\Twisted\twisted\internet\_pollingfile.py", line 72, in _pollEvent
    workUnits += resource.checkWork()
  File "c:\twistedbuildbot24\WXP32-full2.5-scmikes-select\Twisted\twisted\internet\_pollingfile.py", line 129, in checkWork
    self.cleanup()
  File "c:\twistedbuildbot24\WXP32-full2.5-scmikes-select\Twisted\twisted\internet\_pollingfile.py", line 134, in cleanup
    self.lostCallback()
  File "c:\twistedbuildbot24\WXP32-full2.5-scmikes-select\Twisted\twisted\internet\_dumbwin32proc.py", line 289, in outConnectionLost
    self.proto.childConnectionLost(1)
  File "c:\twistedbuildbot24\WXP32-full2.5-scmikes-select\Twisted\twisted\test\test_process.py", line 205, in childConnectionLost
    raise RuntimeError
exceptions.RuntimeError: 
-------------------------------------------------------------------------------
```

<details><summary>Searchable metadata</summary>

```
trac-id__3404 3404
type__defect defect
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__branches_manyprocesses_3404_5 branches/manyprocesses-3404-5
branch_author__exarkun__pahan__glyph exarkun, pahan, glyph
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1220272091000000 1220272091000000
changetime__1413321868384966 1413321868384966
version__None None
owner__glyph glyph

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

(In [24814]) Branching to 'manyprocesses-3404'

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set owner to @exarkun|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set owner to PenguinOfDoom|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="pahan's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|pahan commented|
|-|-|

(In [29116]) Branching to 'manyprocesses-3404-2'

**Comment by twisted-trac** [comment]
|<img alt="pahan's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|pahan commented|
|-|-|

(In [30968]) Branching to 'manyprocesses-3404-3'

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_6"></a>|@glyph commented|
|-|-|

Test logs [like this one](http://buildbot.twistedmatrix.com/builders/winxp32-py2.7/builds/3780/steps/iocp/logs/problems) now show us, thanks to [ticket:5652 a change that added more diagnostic information], that the problem is:

```
exceptions.RuntimeError: Data was '' instead of 'abcd'
```

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_7"></a>|@glyph commented|
|-|-|

Given that some other tests successfully send data back and forth to subprocesses on winxp, it seems that the problem here is that there's a race condition in XP itself where a process can write some data to a pipe, flush it, exit, and then the reader of that pipe doesn't see the data.

This strikes me as a kernel bug.  The fact that I don't think I've ever seen this particular error on another platform is a strong indication.

I could work around it by waiting for a response in the subprocess before exiting, I guess; I'm not sure if that changes what the test tests.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set owner to @glyph<br>@glyph set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_9"></a>|@glyph commented|
|-|-|

(In [43294]) Branching to manyprocesses-3404-4.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_10"></a>|@glyph commented|
|-|-|

I added some more telemetry to try to figure out if there were any interesting boundary conditions.

I think that 50 concurrent subprocesses may just be more than Windows XP can handle.  The test is basically valid ("can we spawn a bunch of processes and get notifications from all of them"), there are no race conditions I can see in the code that reads from the various pipes; just, sometimes, the pipe disappears before all the bytes exist in it.

One of the bits of telemetry I put in was a thing that indicated _which_ processes were failing - by which I mean, identifying the spawn order of the processes and reporting which ones have issues.  Regardless of what other debugging or belt-and-suspenders error-checking I put in, only processes 46 and up ever seemed to fail.  When I increased the number of processes, [reliably, more of them failed; almost always all of them after process 55 or so](http://buildbot.twistedmatrix.com/builders/winxp32-py2.7/builds/3827/steps/select/logs/stdio).

I'm inclined to just decrease the number of concurrent processes to 25, since that still stress-tests the concurrency logic, but does so below the boundary that XP is likely to fail on.  The right solution to this reliability issue is (A) don't use Windows XP if reliability with massive concurrency is a significant concern, and (B) we should get rid of polling-based subprocess stuff entirely in favor of something IOCP-based.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_11"></a>|@glyph commented|
|-|-|

(In [43308]) Branching to manyprocesses-3404-5.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_12"></a>|@glyph removed owner<br>@glyph set status to `new`|
|-|-|

[Since this is a reliability issue, I have forced a whole bunch of builds.](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/manyprocesses-3404-5&num_builds=6)

Hopefully the results will be aesthetically pleasing when they're complete.  Right now the xp builder is so backlogged with my debugging attempts that it hasn't had an opportunity to chew on the final branch, but previous indications were positive that this threshold decrease will work.

If we can get this landed before the [upcoming review rally](https://twistedmatrix.com/pipermail/twisted-python/2014-October/028851.html) though, that will mean that reviewers will more than likely get to ignore spurious failures, and our release can be 100% green.  I realize that this is a somewhat disappointing fix for a reliability issue, but this appears to be an intractable resource limitation issue on one of our supported platforms.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_13"></a>|@glyph commented|
|-|-|

Sigh. The builders now appear to also be occasionally failing due to [#6027](https://github.com/twisted/twisted/issues/6027).  Still, more successful than they were.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun commented|
|-|-|

> I'm inclined to just decrease the number of concurrent processes to 25

Why do this in the tests rather than in the code under test?  And/or why do it for all platforms instead of only for Windows XP?

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_15"></a>|@glyph commented|
|-|-|

Replying to [exarkun](#note_14):
> > I'm inclined to just decrease the number of concurrent processes to 25
> 
> Why do this in the tests rather than in the code under test?

This is the smallest possible change to make the test reliable.  I am open to filing an additional ticket to try to get the code under test to be smarter and support concurrency.  I am even open to committing to fix that ticket myself :-).

However, changing the code under test to behave more like a process _pool_ than a process spawner might have uglier, harder-to-predict side effects; it would need to defer spawning new processes until other, existing processes had exited.  And perhaps the limit that we're approaching is affected by subprocesses holding open handles as well, or is account-wide or something like that.  I would want to carefully avoid doing this anywhere but on XP.  Basically, this is a change I understand that could make our test suite significantly more reliable without an appreciable reduction in coverage _today_, but the other changes would require lots more investigation.

> And/or why do it for all platforms instead of only for Windows XP?

25 concurrent processes doesn't seem to provide any less coverage for the supported code paths here than 50 (except for triggering this liminal condition on this one platform), and all things being equal unit tests shouldn't have for loops at all, let alone arbitrarily long loops that slow things down.

(I'm also not just saying that, I squinted real hard at branch coverage reports of `trial -u` on multiple platforms with hundreds of iterations; there's no discernable difference either conceptually or mechanically.)

If there were some kind of actual, inspect-able limit that this test knew it could safely approach - open handles, open file descriptors, hProcess whatevers - I'd like to have some code that detects the limit and then approaches it.  If we had a limit that we knew we could approach though, there'd be some defined failure mode around hitting it and we could test that too, not just watch the platform drop data on the floor.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml set owner to @jml|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_17"></a>|@jml set owner to @glyph|
|-|-|

Please add a brief comment referring to this ticket. e.g.

```
    # This number was not chosen out of thin air. See #3404 for details.
```

And then feel free to merge.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_18"></a>|@glyph set status to `closed`|
|-|-|

(In [43371]) Merge manyprocesses-3404-5: run fewer processes in testManyProcesses

Author: glyph

Reviewer: jml

Fixes: [#3404](https://github.com/twisted/twisted/issues/3404)

This adjusts the number of concurrent processes run by testManyProcesses
to stay under some obscure, undocumented limit on windows XP and
facilitate more reliable build results.  It should also make the tests
slightly faster on other platforms, so we're changing it globally rather
than platform-specifically.
